### PR TITLE
feat: add microbenchmarks for OM CT parsing

### PR DESCRIPTION
**Pull request by Maniktherana**
partially address #14823 

Adds openmetrics with ct parsing enabled to `BenchmarkParse`

cc: @bwplotka @ArthurSens

### CODE REVIEW

File: model/textparse/promparse_test.go [hunk 1]
```diff
@@ -498,8 +498,16 @@ func BenchmarkParse(b *testing.B) {
...
 			return NewOpenMetricsParser(b, st)
 		},
+		"openmetrics/skip-ct": func(b []byte, st *labels.SymbolTable) Parser {
+			return NewOpenMetricsParser(b, st, WithOMParserCTSeriesSkipped())
+		},
 	} {
-		for _, fn := range []string{"promtestdata.txt", "promtestdata.nometa.txt"} {
+		for _, fn := range []string{"promtestdata.txt", "promtestdata.nometa.txt", "omtestdata.txt"} {
+			// we only want to benchmark om parsers with omtestdata
+			if fn == "omtestdata.txt" && parserName == "prometheus" {
```
Comment (ArthurSens) [inline model/textparse/promparse_test.go:507]:
```suggestion
			if fn == "omtestdata.txt" && !strings.HasPrefix(fn, "openmetrics") {
```

I think we could make this more future-proof in case we add more things here 🤔

File: model/textparse/promparse_test.go [hunk 2]
```diff
@@ -631,6 +639,35 @@ func BenchmarkParse(b *testing.B) {
...
+					for i < b.N {
+						t, err := p.Next()
+						switch t {
+						case EntryInvalid:
+							if errors.Is(err, io.EOF) {
+								break Outer
+							}
+							b.Fatal(err)
+						case EntrySeries:
+							if parserName == "openmetrics/skip-ct" {
```
Comment (ArthurSens) [inline model/textparse/promparse_test.go:663]: I believe this conditional might make a small difference in the benchmark result, we're doing one extra operation that we probably won't do in the real code.

What do you think of moving this conditional up, before `b.Run(....`? I believe that's what you wanted to do, right? Just benchmark this if we're doing open-metrics

Comment (Maniktherana) [reply model/textparse/promparse_test.go:663]: hmm,
So i wanted both om parsers (with and without skipCT) to go through this test but I only call `p.createdTimestamp()` when skipCT is enabled. I believe thats what the goal is but we can't call that function when skipCT is disabled right?

or do we want to just test `p.createdTimestamp()` for om with skipct enabled?

File: model/textparse/promparse_test.go (hunk 1 above)
Comment (Maniktherana) [reply model/textparse/promparse_test.go:507]: yes this is better!

File: model/textparse/promparse_test.go (hunk 2 above)
Comment (Maniktherana) [reply model/textparse/promparse_test.go]: gonna close this one and restart conversation below

**Comment by Maniktherana** [comment]
Still unsure about approach here
I believe it makes sense that I keep benchmarking for `CreatedTimestamp()` function separate from `BenchmarkParse` since the former only benchmark's a specific portion of the parser

But I can still add the OM parser with skipCT enabled in `BenchmarkParse` function

**Comment by ArthurSens** [comment]
I've spent some time looking at the benchmarks we have today and it seems like we benchmark both parsers with Prometheus-text exposition format. See how the test data doesn't have `_created` lines, doesn't have `# UNIT` lines, etc. It's fine benchmarking both Prometheus and OpenMetrics parsers since the Prometheus text format is a subset of OpenMetrics.

Fundamentally what we want to benchmark here is things that differentiate OM from Prometheus format, so I think it make sense to write a separate Benchmark function.

What do you think of renaming `BenchmarkParse` to `BenchmarkPrometheusTxtParse` and then writing a `BenchmarkOpenMetricsTxtParse`? The new benchmark will not use the current testdata, and the old benchmarks will not use the new testdata, this way we avoid a lot of conditionals that you had to add at the moment

**Comment by ArthurSens** [comment]
> Sounds good to me
> 
> will `BenchmarkPrometheusTxtParse` have both OM parsers (CT enabled and disabled)? or is that just for `BenchmarkOpenMetricsTxtParse`

I would keep only the default behavior (not skipping CT) for PrometheusTxtParse.

**Comment by Maniktherana** [comment]
> > Sounds good to me
> > will `BenchmarkPrometheusTxtParse` have both OM parsers (CT enabled and disabled)? or is that just for `BenchmarkOpenMetricsTxtParse`
> 
> I would keep only the default behavior (not skipping CT) for PrometheusTxtParse.

pushed the changes

File: model/textparse/openmetricsparse_test.go [hunk 3]
```diff
@@ -992,3 +994,52 @@ go_gc_duration_seconds_created`)
...
+		require.NoError(b, err)
+		defer f.Close()
+
+		buf, err := io.ReadAll(f)
+		require.NoError(b, err)
+
+		b.Run(parserName+"/parse-ct/"+"omtestdata.txt", func(b *testing.B) {
+			if !strings.HasPrefix(parserName, "openmetrics") {
+				b.Skip()
+			}
```
Comment (ArthurSens) [inline model/textparse/openmetricsparse_test.go:1017]: We don't need this now that this benchmark only covers OM, right?

**Review by ArthurSens — APPROVED** [review]
Small nit, but LGTM!

@bwplotka could you take a look as well?

File: model/textparse/openmetricsparse_test.go (hunk 3 above)
Comment (Maniktherana) [reply model/textparse/openmetricsparse_test.go:1017]: Could've sworn I got rid of this 😵

File: model/textparse/openmetricsparse_test.go [hunk 4]
```diff
@@ -992,3 +994,52 @@ go_gc_duration_seconds_created`)
...
+			b.SetBytes(int64(len(buf) / promtestdataSampleCount))
+			b.ReportAllocs()
+			b.ResetTimer()
+
+			total := 0
```
Comment (ArthurSens) [inline model/textparse/openmetricsparse_test.go:1018]: actually, let me ask one thing.

I noticed that you copied this part from the previous benchmark, but do we know what `total` is used for? It's not obvious to me at first glance, maybe not needed here?

File: model/textparse/openmetricsparse_test.go (hunk 3 above)
Comment (Maniktherana) [reply model/textparse/openmetricsparse_test.go]: Hmm looks like wasn't pushed for some reason

File: model/textparse/openmetricsparse_test.go (hunk 4 above)
Comment (Maniktherana) [reply model/textparse/openmetricsparse_test.go:1018]: I actually am curious as to what its used for 
but yeah don't need this here

Comment (Maniktherana) [reply model/textparse/openmetricsparse_test.go]: removed now

File: model/textparse/promparse_test.go [hunk 5]
```diff
@@ -492,7 +492,7 @@ const (
 	promtestdataSampleCount = 410
 )
 
-func BenchmarkParse(b *testing.B) {
+func BenchmarkPromParse(b *testing.B) {
```
Comment (bwplotka) [inline model/textparse/promparse_test.go:495]: Hm, why this change? Are we not benchmark the openmetrics case as well here? 🤔

Comment (Maniktherana) [reply model/textparse/promparse_test.go:495]: We are but made this distinction because there's a `BenchmarkParse` in openmetrics test too

Rational here is  Prom and OM here are to indicate the input text

File: model/textparse/promparse_test.go [hunk 6]
```diff
@@ -492,7 +492,7 @@ const (
...
 	for parserName, parser := range map[string]func([]byte, *labels.SymbolTable) Parser{
 		"prometheus": NewPromParser,
 		"openmetrics": func(b []byte, st *labels.SymbolTable) Parser {
```
Comment (bwplotka) [inline model/textparse/promparse_test.go:498]: Does this even makes sense? Why we use openmetrics parser for `"promtestdata.txt", "promtestdata.nometa.txt"`?

File: model/textparse/openmetricsparse_test.go [hunk 7]
```diff
@@ -992,3 +993,46 @@ go_gc_duration_seconds_created`)
...
+		},
+	} {
+		f, err := os.Open("omtestdata.txt")
+		require.NoError(b, err)
+		defer f.Close()
+
+		buf, err := io.ReadAll(f)
+		require.NoError(b, err)
+
+		b.Run(parserName+"/parse-ct/"+"omtestdata.txt", func(b *testing.B) {
```
Comment (bwplotka) [inline model/textparse/openmetricsparse_test.go:1013]: Is this test really different than `BenchmarkParse`? can we create one that makes sense?

File: model/textparse/promparse_test.go (hunk 6 above)
Comment (Maniktherana) [reply model/textparse/promparse_test.go:498]: 🤷
OM parser was already here but in a comment above Arthur mentioned that Prom text is a subset of OM so it works out

File: model/textparse/openmetricsparse_test.go (hunk 7 above)
Comment (Maniktherana) [reply model/textparse/openmetricsparse_test.go:1013]: In this one we're benchmarking just the performance of calling p.CreatedTimestamp on each line of a "scrape". This was my understanding of the requirements

Is there anything else I should consider?

**Review by bwplotka — CHANGES_REQUESTED** [review]
Nice!

I wonder if it would make sense to actually clean `BenchmarkParse`. Perhaps put it in separate `benchmark_test.go` file, and have a clear 3 cases that always call Created Timestamp (with https://www.bwplotka.dev/2024/go-microbenchmarks-benchstat/ case naming):

```
for _, in := range []string{"omtestdata.txt", "promtestdata.txt", "promtestdata.nometa.txt"} {
b.Run(fmt.Sprintf("input=%v"), in, func....)
}
```

Then chose the parser based on input format. 

Does it make sense to have benchmark for skip true and false? Is this really that different between those? I think the main problem is CreatedTimestamp implementation right now, no matter if we skip or not, and I would capture the benchmark for default path (skip = true). We can do adhoc benchmark run with skip = false for testing, but not need to capture this perhaps?

File: model/textparse/openmetricsparse_test.go (hunk 7 above)
Comment (bwplotka) [reply model/textparse/openmetricsparse_test.go:1013]: I mean it would be nice to have one benchmark for all of those, for consistency. Why we benchmark Created Timestamp for OM but not for Prometheus proto?

**Comment by Maniktherana** [comment]
> Does it make sense to have benchmark for skip true and false? Is this really that different between those? I think the main problem is CreatedTimestamp implementation right now, no matter if we skip or not, and I would capture the benchmark for default path (skip = true). We can do adhoc benchmark run with skip = false for testing, but not need to capture this perhaps?

Yes this makes sense. I'll make some changes

Although I still feel we can keep a benchmarkParse in omparse and promparse based off which parser we use
(But that's just me and I could be wrong)

File: model/textparse/promparse_test.go (hunk 6 above)
Comment (bwplotka) [reply model/textparse/promparse_test.go:498]: Interesting. Is this something that users are doing (parsing PromText with OpenMetrics text)?

**Comment by bwplotka** [comment]
> Although I still feel we can keep a benchmarkParse in omparse and promparse based off which parser we use
(But that's just me and I could be wrong)

Is this duplication of benchmark code (that ages quickly) worth it? Why?

File: model/textparse/promparse_test.go (hunk 5 above)
Comment (bwplotka) [reply model/textparse/promparse_test.go:495]: Got it, thanks! It looks like the benchmark code is generally duplicated though 🤔

File: model/textparse/promparse_test.go (hunk 6 above)
Comment (bwplotka) [reply model/textparse/promparse_test.go:498]: I guess they were or someone wanted to compare it, ack! thx

File: model/textparse/openmetricsparse_test.go (hunk 7 above)
Comment (bwplotka) [reply model/textparse/openmetricsparse_test.go:1013]: .. but not a biggie for this PR

File: model/textparse/openmetricsparse_test.go [hunk 8]
```diff
@@ -992,3 +993,46 @@ go_gc_duration_seconds_created`)
 	require.Equal(t, "go_gc_duration_seconds", string(copyParser.l.b[copyParser.offsets[0]:copyParser.offsets[1]]))
 	require.False(t, copyParser.skipCTSeries)
 }
+
+func BenchmarkOMParse(b *testing.B) {
```
Comment (bwplotka) [inline model/textparse/openmetricsparse_test.go:997]:
```suggestion
func BenchmarkOMParseCreatedTimestamp(b *testing.B) {
```

**Review by bwplotka — DISMISSED** [review]
Actually, let's do it like that, just name the omtext benchmark a bit more explicitly (: We can improve this one later.

Thanks!  LGTM

File: model/textparse/openmetricsparse_test.go [hunk 9]
```diff
@@ -992,3 +993,46 @@ go_gc_duration_seconds_created`)
...
+	for parserName, parser := range map[string]func([]byte, *labels.SymbolTable) Parser{
+		"openmetrics": func(b []byte, st *labels.SymbolTable) Parser {
+			return NewOpenMetricsParser(b, st)
+		},
+		"openmetrics-skip-ct": func(b []byte, st *labels.SymbolTable) Parser {
```
Comment (bwplotka) [inline model/textparse/openmetricsparse_test.go:1002]: Perhaps it would make sense here to always do openmetrics (no casing)but haveb.Run(fmt.Sprintf("skip-ct=%v", skipCT), ....)` case, WDYT?

Comment (Maniktherana) [reply model/textparse/openmetricsparse_test.go:1002]: yes I agree with that bit
virtually no difference here for both parsers

Comment (bwplotka) [reply model/textparse/openmetricsparse_test.go:1002]: Happy to do in separare PR

**Review by bwplotka — APPROVED** [review]
